File: csharp/LoanServer/LoanHandler.cs
using System;
using System.IO;
using System.Net;

namespace MikadoMethod.LoanServer {
  public class LoanHandler {

    public const string APPLICATION = "apply";
    public const string FETCH = "fetch";
    public const string TICKET_ID = "ticketId";
    public const string APPROVE = "approve";

    public void HandleRequest(HttpListenerContext context) {

      var request = context.Request;
      var response = context.Response;
      var writer = new StreamWriter(response.OutputStream);

      response.ContentType = "application/json;charset=utf-8";

      try {
        if(IsApplication(request)) {
          LoanApplication application = new LoanApplication();
          application.Amount = AmountFrom(request);
          application.Contact = ContactFrom(request);
          Ticket ticket = LoanRepository.Store(application);
          writer.WriteLine(Json.To<Ticket>(ticket));
        } else if(IsStatusRequest(request) && IdSpecified(request)) {
          writer.WriteLine(FetchLoanInfo(request.QueryString.Get(TICKET_ID)));
        } else if(IsApproval(request) && IdSpecified(request)) {
          writer.WriteLine(ApproveLoan(request.QueryString.Get(TICKET_ID)));
        } else {
          response.StatusCode = (int)HttpStatusCode.BadRequest;
          writer.WriteLine("Incorrect parameters provided");
        }
      } catch(Exception e) {
        response.StatusCode = (int)HttpStatusCode.BadRequest;
        writer.WriteLine("Uh oh! Problem occured: {0}", e.Message);
      }

      writer.Close();
    }

    private string ContactFrom(HttpListenerRequest request) {
      return request.QueryString.Get("contact");
    }

    private long AmountFrom(HttpListenerRequest request) {
      return long.Parse(request.QueryString.Get("amount"));
    }

    private string ApproveLoan(string parameter) {
      return Json.To<Ticket>(LoanRepository.Approve(parameter));
    }

    private bool IsApproval(HttpListenerRequest request) {
      return APPROVE.Equals(request.QueryString.Get("action"));
    }

    private bool IdSpecified(HttpListenerRequest request) {
      return request.QueryString.Get(TICKET_ID) != null && ValidId(request) >= 0;
    }

    private long ValidId(HttpListenerRequest request) {
      long ticketId;
      if (long.TryParse(request.QueryString.Get(TICKET_ID), out ticketId)) {
        return ticketId;
      }
      return -1;
    }

    private bool IsStatusRequest(HttpListenerRequest request) {
      return FETCH.Equals(request.QueryString.Get("action"));
    }

    private bool IsApplication(HttpListenerRequest request) {
      return APPLICATION.Equals(request.QueryString.Get("action"));
    }

    private string FetchLoanInfo(string ticketId) {
      return Json.To<LoanApplication>(LoanRepository.Fetch(ticketId));
    }

    public static long GetNextId() {
      try {
        var files = Directory.GetFiles(LoanRepository.REPOSITORY_ROOT, "*" + LoanRepository.FILE_EXTENSION);
        return files.Length;
      } catch(DirectoryNotFoundException) {
        return 0;
      }
    }

  }
}

File: cpp/loan_server/source/LoanHandler.cc
#include <loan_server/LoanHandler.h>

#include <loan_server/LoanApplication.h>
#include <loan_server/LoanRepository.h>
#include <loan_server/Ticket.h>

#include <fstream>

namespace
{
  std::string const APPLICATION = "apply";
  std::string const FETCH = "fetch";
  std::string const TICKET_ID = "ticketId";
  std::string const APPROVE = "approve";

  std::string contact_from(http::Request const & request) {
    return request.parameters.at("contact");
  }

  long amount_from(http::Request const & request) {
    return stol(request.parameters.at("amount"));
  }

  std::string approve_loan(std::string parameter) {
    auto const json =Json{mikado::LoanRepository::approve(parameter)};
    return json.dump();
  }

  bool is_approval(http::Request const & request) {
    try {
      return APPROVE == request.parameters.at("action");
    } catch(std::out_of_range const &) {
      return false;
    }
  }

  long valid_id(http::Request const & request) {
    try {
      return stol(request.parameters.at(TICKET_ID));
    } catch(std::invalid_argument const &) {
      return -1;
    }
  }

  bool id_specified(http::Request const & request) {
    return (request.parameters.find(TICKET_ID) != request.parameters.end())
      && (valid_id(request) >= 0);
  }

  bool is_status_request(http::Request const & request) {
    try {
      return FETCH == request.parameters.at("action");
    } catch(std::out_of_range const &) {
      return false;
    }
  }

  bool is_application(http::Request const & request) {
    try {
      return APPLICATION == request.parameters.at("action");
    } catch(std::out_of_range const &) {
      return false;
    }
  }

  std::string fetch_loan_info(std::string ticket_id) {
    auto const json = Json{mikado::LoanRepository::fetch(ticket_id)};
    return json.dump();
  }
}

namespace mikado
{
  http::Response LoanHandler::handle(http::Request request) const {
    http::Response response{};
    response.content_type = "Content-Type: application/json;charset=utf-8";
    response.status_code = 200;

    try {
      if(is_application(request)) {
        LoanApplication application{LoanApplication::new_application()};
        application.amount = amount_from(request);
        application.contact = contact_from(request);
        auto const ticket = LoanRepository::store(application);
        response.response = Json{ticket}.dump();
      } else if(is_status_request(request) && id_specified(request)) {
        response.response = fetch_loan_info(request.parameters.at(TICKET_ID));
      } else if(is_approval(request) && id_specified(request)) {
        response.response = approve_loan(request.parameters.at(TICKET_ID));
      } else {
        Json error = {{"error", "Incorrect parameters provided"}};
        response.response = error.dump();
        response.status_code = 400;
      }
    } catch(std::exception const & e) {
      Json error = {{ "error", "Uh oh! Problem occured: " + std::string{e.what()} }};
      response.response = error.dump();
      response.status_code = 500;
    }

    return response;
  }

  long LoanHandler::get_next_id() {
    std::ifstream read_id(LoanRepository::REPOSITORY_ROOT + "/last.id", std::ios::binary);
    long last_id = 0;
    read_id >> last_id;
    read_id.close();

    long next_id = ++last_id;

    std::ofstream write_id(LoanRepository::REPOSITORY_ROOT + "/last.id", std::ios::binary);
    write_id << next_id;
    write_id.close();

    return next_id;
  }
}
